langchain/guardrails/parsing.py

```diff
@@ -0,0 +1,30 @@
+from typing import Any
+
+from langchain.guardrails.retry import naive_retry
+from langchain.output_parsers import BaseOutputParser, OutputParserException
+from langchain.schema import BaseLanguageModel, PromptValue
+
+
+class RetriableOutputParser(BaseOutputParser):
+    """Wraps a parser and tries to fix parsing errors."""
+
+    parser: BaseOutputParser
+    retry_llm: BaseLanguageModel
+
+    def parse_with_prompt(self, completion: str, prompt_value: PromptValue) -> Any:
+        try:
+            parsed_completion = self.parser.parse(completion)
+        except OutputParserException as e:
+            # TODO: can also inject str(e) into prompt, but have to validate this as a reasonable default.
+            new_completion = naive_retry(
+                llm=self.retry_llm, prompt=prompt_value, completion=completion,
+            )
+            parsed_completion = self.parser.parse(new_completion)
+        
+        return parsed_completion
+
+    def parse(self, completion: str):
+        raise NotImplementedError
+
+    def get_format_instructions(self) -> str:
+        return self.parser.get_format_instructions()
```

Comment:
@hwchase17 we wanted a "guarded" parser to itself be a parser so it can be swappable at the call site. to do this, BaseOutputParser.parse_with_prompt invokes BaseOutputParser.parse but can be overridden, like here.   however, conceptually guardrail/retry is just a wrapper... we'd like to be able to use this with any of the parser. so this both "is a" and "has a"